### PR TITLE
only keep SNTemp outlets in distance_matrix

### LINKED ISSUE #19: Mismatch in adjacency matrix/ sntemp output full DRB
There is a mismatch in the number of segments in the adjacency matrix and the sntemp output for the full DRB.

The adjacency matrix (i.e. updown.npy in the UMN-USGS Data Sharing/Streams GCN/Full Network/NN Inputs/dists_full.npz) is 459 x 459.

In the sntemp_input_output.feather (from the same GDrive folder), there are 456 segments.

(This was originally an issue in my ML model repo, so I will transfer comments from there.)

### LINKED ISSUE #20: Order of seg_id's for distance matrix
I'm wondering what order the seg_id's are in when the distance matrix is calculated. In other words, which row in the distance matrix corresponds to which seg_id. 

I don't think the order itself will matter, but I think it will make a difference if the order is *different* in the input datasets to the RGCN model compared to the order in the distance matrix.

### PR DESCRIPTION
**Pull request by aappling-usgs**
Resolves #19 by subsetting the distance matrix (9_collaborator_data/umn/distance_matrix.npz) to only those segments that end at the SNTemp segment outlets. Also resolves #20 by adding an item to the .npz with the subseg_ids in the order that they appear in the distance matrices (top to bottom, left to right)

I had to rebuild a bunch of stuff. I tried figuring out why scipiper wanted to rebuild so much but didn't get there. I did rebuild every target, I think, which I hope might help resolve any inconsistencies within the pipeline if those existed. There were a couple of targets that rebuilt even after I'd just built them, which is odd and not something I can explain yet: 9_collaborator_data/psu/highly_obs_distance_matrix.csv.ind and 9_collaborator_data/umn/distance_matrix.npz.ind.

### CODE REVIEW

**Comment by aappling-usgs** [comment]
1_network/out/network.rds has the mapping from subseg_id to seg_id_nat

**Comment by jsadler2** [comment]
This may not be relevant to the actual R code that you added, but I just pulled your branch and ran `which_dirty` and I think it wants me to rebuild everything

```R
> scipiper::which_dirty(remake_file = 'remake.yml')
 [1] "1_network/out/distance_matrix.rds.ind"                           "1_network"                                                      
 [3] "2_observations/out/crosswalk_site_reach.rds.ind"                 "2_observations/out/drb_filtered_sites.rds.ind"                  
 [5] "2_observations/out/all_drb_temp_obs.rds.ind"                     "2_observations/out/obs_temp_drb.rds.ind"                        
 [7] "2_observations/out/obs_flow_drb.rds.ind"                         "2_observations"                                                 
 [9] "8_visualize/out/site_obscount_freq_dist.rds"                     "8_visualize/out/site_obscount_freq_dist.png"                    
[11] "8_visualize/out/nreaches_nobs_nyears.png"                        "9_collaborator_data/mkrsp/delaware_sites_summary.rds.ind"       
[13] "9_collaborator_data/mkrsp/delaware_sites_summary.geojson.ind"    "9_collaborator_data/umn/sntemp_segs.rds"                        
[15] "9_collaborator_data/umn/sntemp_distance_matrix.rds.ind"          "9_collaborator_data/umn/distance_matrix.npz.ind"                
[17] "9_collaborator_data/umn/network_subset.rds.ind"                  "8_visualize/out/map_network_subset.png"                         
[19] "8_visualize"                                                     "9_collaborator_data/umn/network_subset_sntemp_segs.rds"         
[21] "9_collaborator_data/umn/distance_matrix_subset.rds.ind"          "9_collaborator_data/umn/distance_matrix_subset.npz.ind"         
[23] "9_collaborator_data/umn/obs_temp_full.csv.ind"                   "9_collaborator_data/umn/obs_temp_subset.csv.ind"                
[25] "9_collaborator_data/umn/obs_flow_full.csv.ind"                   "9_collaborator_data/umn/obs_flow_subset.csv.ind"                
[27] "9_collaborator_data/psu/highly_observed_sites.rds"               "9_collaborator_data/psu/highly_observed_sites.csv.ind"          
[29] "9_collaborator_data/psu/highly_obs_distance_matrix.rds.ind"      "9_collaborator_data/psu/highly_obs_distance_matrix.rds"         
[31] "9_collaborator_data/psu/highly_obs_distance_matrix.csv.ind"      "9_collaborator_data/psu/highly_obs_distance_heatmap.png"        
[33] "9_collaborator_data/psu/high_obs_upstream_sites.csv"             "9_collaborator_data/psu/sntemp_preds_high_obs_sites.feather.ind"
[35] "9_collaborator_data/psu/sntemp_preds_high_obs_sites.feather"     "9_collaborator_data/psu/sntemp_preds_aggregated.feather.ind"    
[37] "9_collaborator_data/psu/temp_obs_high_obs_sites.csv.ind"         "9_collaborator_data/psu/flow_obs_high_obs_sites.csv.ind"        
[39] "9_collaborator_data"                                             "3_predictions"                                                  
[41] "everything"                                                     
```

**Comment by aappling-usgs** [comment]
Sigh. Can't say I'm surprised =(. Please don't bother rerunning for now, especially because I have two open PRs at the moment.

**Comment by jsadler2** [comment]
Sounds good. I'll wait on rerunning. Did this only start happening when I started using the pipeline? Is it something with my setup?

**Comment by aappling-usgs** [comment]
If I knew for sure, we wouldn't still be having problems 😝 . But I don't think it's just your setup because my computer wanted to rebuild lots of things after getting a hand-off from Sam a few weeks ago.

File: 9_collaborator_data/src/psu_site_selection.R [hunk 1]
```diff
@@ -315,33 +315,47 @@ plot_upstream <- function(POIs, dist_mat_ind, labels=c('subseg_id','seg_id_nat')
...
+  saveRDS(sntemp_edges, out_rds)
+}
 
-  subset_dist <- dist_mat[sites$subseg_id, ]
-  subset_dist <- subset_dist[, sites$subseg_id]
+subset_dist_to_subsegs <- function(subsegs_rds, dist_ind, network_ind, out_ind){
 
-  saveRDS(subset_dist, as_data_file(out_ind))
+  subsegs <- readRDS(subsegs_rds)
+  network <- readRDS(sc_retrieve(network_ind))
```
Comment (jsadler2) [inline 9_collaborator_data/src/psu_site_selection.R]: It doesn't look like you use this variable.

Comment (jsadler2) [reply 9_collaborator_data/src/psu_site_selection.R]: (i.e., `network`) so I think you can remove it and the function arg

File: 9_collaborator_data/src/psu_site_selection.R [hunk 2]
```diff
@@ -315,33 +315,47 @@ plot_upstream <- function(POIs, dist_mat_ind, labels=c('subseg_id','seg_id_nat')
...
+  dist_mats <- readRDS(sc_retrieve(dist_ind))
+  subset_dists <- lapply(dist_mats, function(dist_mat) {
```
Comment (jsadler2) [inline 9_collaborator_data/src/psu_site_selection.R]: question: what is `dist_mat`? I don't see it defined anywhere.

Comment (jsadler2) [reply 9_collaborator_data/src/psu_site_selection.R]: oh nvm. it's a temporary function you are defining here.

Comment (jsadler2) [reply 9_collaborator_data/src/psu_site_selection.R]: jk again 😆 . It's the argument to the function you are defining here - the variable represents the members of the list

File: 9_collaborator_data/src/psu_site_selection.R [hunk 3]
```diff
@@ -315,33 +315,47 @@ plot_upstream <- function(POIs, dist_mat_ind, labels=c('subseg_id','seg_id_nat')
...
+    subset_dist <- dist_mat[subsegs$subseg_id, ][, subsegs$subseg_id]
```
Comment (jsadler2) [inline 9_collaborator_data/src/psu_site_selection.R]: It would be more helpful if the column and row names were the `seg_id_nat`s instead of the `subseg_id`s since the sntemp output has the former and not the latter. I think we could just add two lines here:

```R
colnames(subset_dist) <- colnames(subsegs$seg_id_nat)
rownames(subset_dist) <- colnames(subsegs$seg_id_nat)
```
But maybe this should be its own PR?

**Comment by jsadler2** [comment]
Also, I manually downloaded `distance_matrix.npz` and confirmed that it is the expected shape (456x456). I still can't easily match the order to the sntemp output b/c sntemp has `seg_id_nat` but not `subseg_id`. I suggest making the matrix colnames and rownames the `seg_id_nat`s [here](https://github.com/USGS-R/delaware-model-prep/pull/21#discussion_r400488885)

File: 9_collaborator_data.yml [hunk 4]
```diff
@@ -23,6 +23,7 @@ targets:
     depends:
      - 9_collaborator_data/mkrsp/delaware_sites_summary.geojson.ind
      - 9_collaborator_data/umn/network_subset.rds.ind
+     - 9_collaborator_data/umn/sntemp_distance_matrix.rds.ind
```
[the remark this answers is not in the document]
Comment (aappling-usgs) [reply 9_collaborator_data.yml]: Good point. To distinguish from the distance matrix based on subsegments in 1_network, I'll rename that earlier target to subseg_distance_matrix and then make both of these just distance_matrix.

[the remark this answers is not in the document]
Comment (aappling-usgs) [reply 9_collaborator_data.yml]: The .rds version already existed before I started working with this code; I was just adding it to the list of targets at the top. Yes, I believe it was for convenience, though I think if I were rewriting this pipeline in full I would probably only bother to designate the .npz as a target worth declaring and pushing to Drive.

[the remark this answers is not in the document]
Comment (aappling-usgs) [reply 9_collaborator_data.yml]: OK, @jsadler2 , I've added commits to fix this one.

File: 9_collaborator_data/src/psu_site_selection.R [hunk 5]
```diff
@@ -315,33 +315,47 @@ plot_upstream <- function(POIs, dist_mat_ind, labels=c('subseg_id','seg_id_nat')
 
 }
 
+# identify those subsegs whose outlets are in the geospatial fabric (omitting
+# reaches we added for the purpose of constructing and accurate distance matrix)
+select_sntemp_subsegs <- function(
```
[the remark this answers is not in the document]
Comment (aappling-usgs) [reply 9_collaborator_data/src/psu_site_selection.R]: Good call. I've been wrestling with hourly confusion about which project I'm working on at that moment =)

File: 9_collaborator_data/src/psu_site_selection.R (hunk 3 above)
Comment (aappling-usgs) [reply 9_collaborator_data/src/psu_site_selection.R]: No, I agree with doing this in this PR, within `subset_dist_to_subsegs()`.
